File: java/src/main/java/br/com/emendes/problems/RemoveDuplicatesFromSortedList.java
package br.com.emendes.problems;

import br.com.emendes.problems.util.ListNode;

/**
 * Dado o head de uma lista encadeada ordenada,
 * delete todas as duplicatas de maneira que cada elemento apareça uma vez.
 * Retorne a lista encadeada ordenada.
 * <br><br>
 * Restrições:<br>
 * O número de nodes da lista está no intervalo [0, 300].<br>
 * -100 <= Node.val <= 100<br>
 * É garantido que a lista estará ordenada.
 */
public class RemoveDuplicatesFromSortedList {

  public ListNode deleteDuplicates(ListNode head) {
    ListNode dummy = new ListNode(0, head);

    ListNode curr = head;
    ListNode previous = dummy;

    while (curr != null) {
      while (curr.next != null && curr.val == curr.next.val) {
        curr = curr.next;
      }
      if (previous.next == curr) {
        previous = previous.next;
      } else {
        previous.next = curr;
        previous = curr;
      }
      curr = curr.next;
    }

    return dummy.next;
  }

}

File: javascript/problems/RemoveDuplicatesFromSortedList.js
/**
 * Dado o head de uma lista encadeada, delete todas as duplicatas de modo que
 * cada elemento apareça apenas uma vez. Retorne a lista ordenada.
 *
 * Restrições:
 *
 * o número de nodes está no intervalo [0, 300].
 *
 * -100 <= Node.val <= 100
 *
 * É garantido que a lista esta ordenada.
 */
const deleteDuplicates = function (head) {
  let dummy = new ListNode(-1, head);
  while (head) {
    if (head.val == head.next?.val) head.next = head.next.next;
    else head = head.next;
  }

  return dummy.next;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

function generateListNode(values) {
  const dummy = new ListNode(-1);
  let node = dummy;

  for (const val of values) {
    node.next = new ListNode(val);
    node = node.next;
  }
  return dummy.next;
}
